airstack-modules/modules/airstack/domain-name/domain-name.ts

```diff
@@ -0,0 +1,1074 @@
+      entity.registrationCost = BIG_INT_ZERO;
+      entity.createdAt = domain.block.id;
+      entity.lastBlock = domain.block.id;
+    }
+    if (entity.id == ROOT_NODE) {
+      entity.isMigrated = true;
+    }
+    entity.save();
+    return entity as AirDomain;
+  }
+
+  /**
+   * @dev this function gets or creates a new air domain owner changed transaction entity
+   * @param block air block entity
+   * @param logIndex log index
+   * @param chainId chain id
+   * @param previousOwnerId previous owner id
+   * @param newOwner new owner address
+   * @param transactionHash transaction hash
+   * @param tokenId token id
+   * @param domain air domain
+   * @returns AirDomainOwnerChangedTransaction entity
+   */
+  function getOrCreateAirDomainOwnerChangedTransaction(
+    block: AirBlock,
+    logIndex: BigInt,
+    chainId: string,
+    previousOwnerId: string,
+    newOwner: string,
+    transactionHash: Bytes,
+    tokenId: string,
+    domain: AirDomain,
+  ): AirDomainOwnerChangedTransaction {
+    let id = createEntityId(transactionHash, block.number, logIndex);
+    let entity = AirDomainOwnerChangedTransaction.load(id);
+    if (entity == null) {
+      entity = new AirDomainOwnerChangedTransaction(id);
+      entity.previousOwner = previousOwnerId;
+      entity.newOwner = getOrCreateAirAccount(chainId, newOwner).id;
+      entity.transactionHash = transactionHash.toHex();
+      entity.tokenId = tokenId;
+      entity.domain = domain.id;
+      entity.block = block.id;
+      entity.index = updateAirEntityCounter(AIR_DOMAIN_OWNER_CHANGED_ENTITY_COUNTER_ID, block);
+      entity.save();
+    }
+    return entity as AirDomainOwnerChangedTransaction;
+  }
+
+  /**
+   * @dev this function does a recursive domain deletion for a particular domain
+   * @param domain air domain entity
+   * @param chainId chain id
+   * @param block air block entity
+   * @param tokenAddress contract address of nft token
+   * @returns parent domain id
+   */
+  function recurseDomainDelete(domain: AirDomain, chainId: string, block: AirBlock, tokenAddress: string): string | null {
+    if (domain.owner == getOrCreateAirAccount(chainId, ZERO_ADDRESS).id && domain.subdomainCount == BIG_INT_ZERO) {
+      if (domain.parent) {
+        const parentDomain = getOrCreateAirDomain(new Domain(domain.parent!, chainId, block, tokenAddress));
+        if (parentDomain) {
+          parentDomain.subdomainCount = parentDomain.subdomainCount.minus(BIGINT_ONE)
+          parentDomain.lastBlock = block.id;
+          parentDomain.save();
+          return recurseDomainDelete(parentDomain, chainId, block, tokenAddress)
+        }
+      }
+      return null
+    }
+    return domain.id
+  }
+
+  /**
+   * @dev this function updates air entity counter for a given entity id
+   * @param id entity id for entity to be updated
+   * @param block air block object
+   * @returns updated entity count
+   */
+  function updateAirEntityCounter(
```

Comment:
Move it to a common file.

Reply:
good idea, but this would also expose the function for people to use outside the module and they could potentially use it to manipulate entity counters.

Reply:
but, they can do it without these functions also, as they will have full access to these entities eitherways, so, I am making this change.
DONE